File: src/services/ip.rs
use anyhow::{bail, Result};
use std::collections::{HashMap, HashSet};
use std::path::PathBuf;
use std::process::Command;

// ── Constants ──

const NETWORK_STATE_FILE: &str = ".tncli/network.json";
const CURRENT_VERSION: u8 = 2;

/// Number of subnets pre-created by `tncli setup`.
pub const SETUP_SUBNET_COUNT: u8 = 10;
/// Max host IPs per subnet created by setup (2..=SETUP_HOST_MAX).
pub const SETUP_HOST_MAX: u8 = 51;

// ── Unified Network State ──

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct NetworkState {
    #[serde(default = "default_version")]
    pub version: u8,
    /// Session name → subnet slot (1-based).
    #[serde(default)]
    pub subnets: HashMap<String, u8>,
    /// Worktree key → allocated IP.
    #[serde(default)]
    pub allocations: HashMap<String, String>,
}

fn default_version() -> u8 { 1 }

impl Default for NetworkState {
    fn default() -> Self {
        Self { version: CURRENT_VERSION, subnets: HashMap::new(), allocations: HashMap::new() }
    }
}

// ── Helpers ──

fn home_path(rel: &str) -> PathBuf {
    let home = std::env::var("HOME").unwrap_or_else(|_| "/tmp".into());
    PathBuf::from(home).join(rel)
}

fn state_path() -> PathBuf { home_path(NETWORK_STATE_FILE) }

pub fn load_network_state() -> NetworkState {
    let path = state_path();
    std::fs::read_to_string(&path)
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or_default()
}

fn save_network_state(state: &NetworkState) {
    let path = state_path();
    if let Some(parent) = path.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    if let Ok(json) = serde_json::to_string_pretty(state) {
        let _ = std::fs::write(&path, json);
    }
}

/// Compat wrapper — callers that read allocations get them from NetworkState.
pub fn load_ip_allocations() -> HashMap<String, String> {
    load_network_state().allocations
}

// ── File Lock ──

/// File-lock protected operation (safe for concurrent pipelines).
pub(crate) fn with_ip_lock<F, R>(f: F) -> R
where
    F: FnOnce() -> R,
{
    use std::fs::OpenOptions;
    use std::io::Write;

    let lock_path = home_path(".tncli/network.lock");
    if let Some(parent) = lock_path.parent() {
        let _ = std::fs::create_dir_all(parent);
    }

    // Spin-lock on file creation (atomic on POSIX)
    let lockfile = loop {
        match OpenOptions::new().write(true).create_new(true).open(&lock_path) {
            Ok(mut f) => { let _ = write!(f, "{}", std::process::id()); break f; }
            Err(_) => {
                // Check if lock is stale (older than 10s)
                if let Ok(meta) = std::fs::metadata(&lock_path) {
                    if let Ok(modified) = meta.modified() {
                        if modified.elapsed().unwrap_or_default() > std::time::Duration::from_secs(10) {
                            let _ = std::fs::remove_file(&lock_path);
                            continue;
                        }
                    }
                }
                std::thread::sleep(std::time::Duration::from_millis(50));
            }
        }
    };

    let result = f();
    drop(lockfile);
    let _ = std::fs::remove_file(&lock_path);
    result
}

// ── Legacy Migration ──

/// One-time migration from v1 (separate files) to v2 (unified network.json).
/// Also clears old 127.0.0.x entries and proxy routes.
/// Idempotent — skipped if network.json already at version 2.
pub fn migrate_legacy_ips() {
    let state = load_network_state();
    if state.version >= CURRENT_VERSION {
        return;
    }

    with_ip_lock(|| {
        let mut state = load_network_state();
        if state.version >= CURRENT_VERSION {
            return; // Another thread migrated
        }

        // Import from old separate files
        let old_loopback = home_path(".tncli/loopback.json");
        if old_loopback.exists() {
            if let Ok(s) = std::fs::read_to_string(&old_loopback) {
                if let Ok(allocs) = serde_json::from_str::<HashMap<String, String>>(&s) {
                    for (k, v) in allocs {
                        if !v.starts_with("127.0.0.") {
                            state.allocations.insert(k, v);
                        }
                    }
                }
            }
            let _ = std::fs::remove_file(&old_loopback);
        }

        let old_subnets = home_path(".tncli/subnets.json");
        if old_subnets.exists() {
            if let Ok(s) = std::fs::read_to_string(&old_subnets) {
                if let Ok(subs) = serde_json::from_str::<HashMap<String, u8>>(&s) {
                    state.subnets = subs;
                }
            }
            let _ = std::fs::remove_file(&old_subnets);
        }

        // Clear legacy 127.0.0.x from allocations
        state.allocations.retain(|_, ip| !ip.starts_with("127.0.0."));

        // Clear proxy routes with old targets
        let mut routes = super::proxy::load_routes();
        let before = routes.routes.len();
        routes.routes.retain(|_, target| !target.starts_with("127.0.0."));
        if routes.routes.len() != before {
            let mut ports: Vec<u16> = routes.routes.keys()
                .filter_map(|k| k.rsplit(':').next()?.parse().ok())
                .collect();
            ports.sort();
            ports.dedup();
            routes.listen_ports = ports;
            super::proxy::save_routes_pub(&routes);
        }

        state.version = CURRENT_VERSION;
        save_network_state(&state);

        // Clean up old migration flag
        let _ = std::fs::remove_file(home_path(".tncli/.migrated-subnet"));
    });
}

// ── Subnet Allocation ──

/// Release a subnet slot for a session.
#[allow(dead_code)]
pub fn release_subnet(session: &str) {
    with_ip_lock(|| {
        let mut state = load_network_state();
        state.subnets.remove(session);
        save_network_state(&state);
    });
}

// ── Loopback IP Allocation ──

/// Get the allocated IP for the main workspace. Allocates one if needed.
pub fn main_ip(session: &str, default_branch: &str) -> String {
    let key = format!("ws-{}", default_branch.replace('/', "-"));
    allocate_ip(session, &key)
}

/// Allocate next available loopback IP within the session's subnet.
/// Format: 127.0.{subnet_slot}.{2..254}
/// Thread-safe via file lock. No sudo — purely file-based.
pub fn allocate_ip(session: &str, worktree_key: &str) -> String {
    // Single lock for both subnet + IP allocation (avoid double-lock deadlock)
    with_ip_lock(|| {
        let mut state = load_network_state();

        // Allocate subnet if needed
        let subnet = if let Some(&slot) = state.subnets.get(session) {
            slot
        } else {
            let used: HashSet<u8> = state.subnets.values().copied().collect();
            let slot = (1..=254u8).find(|n| !used.contains(n)).unwrap_or(254);
            state.subnets.insert(session.to_string(), slot);
            slot
        };

        if let Some(ip) = state.allocations.get(worktree_key) {
            return ip.clone();
        }

        let prefix = format!("127.0.{subnet}.");
        let used: HashSet<&str> = state.allocations.values().map(|s| s.as_str()).collect();
        let mut n = 2u8;
        loop {
            let ip = format!("{prefix}{n}");
            if !used.contains(ip.as_str()) {
                state.allocations.insert(worktree_key.to_string(), ip.clone());
                save_network_state(&state);
                return ip;
            }
            n += 1;
            if n == 255 {
                let fallback = format!("{prefix}254");
                state.allocations.insert(worktree_key.to_string(), fallback.clone());
                save_network_state(&state);
                return fallback;
            }
        }
    })
}

/// Release an allocated IP. Thread-safe via file lock.
pub fn release_ip(worktree_key: &str) {
    with_ip_lock(|| {
        let mut state = load_network_state();
        state.allocations.remove(worktree_key);
        save_network_state(&state);
    });
}

/// Create loopback alias (requires sudo — setup only).
#[allow(dead_code)]
pub fn setup_loopback(ip: &str) -> Result<()> {
    let status = Command::new("sudo")
        .args(["ifconfig", "lo0", "alias", ip])
        .status()?;
    if !status.success() {
        bail!("failed to create loopback alias {ip} (sudo required)");
    }
    Ok(())
}

/// Remove loopback alias.
#[allow(dead_code)]
pub fn teardown_loopback(ip: &str) -> Result<()> {
    let _ = Command::new("sudo")
        .args(["ifconfig", "lo0", "-alias", ip])
        .status();
    Ok(())
}

// ── /etc/hosts ──

pub fn check_etc_hosts(hostnames: &[&str]) -> Vec<String> {
    let content = std::fs::read_to_string("/etc/hosts").unwrap_or_default();
    hostnames.iter()
        .filter(|h| !content.contains(*h))
        .map(|h| h.to_string())
        .collect()
}

/// Add hostnames to /etc/hosts pointing to 127.0.0.1 (setup only — uses sudo).
#[allow(dead_code)]
pub fn setup_etc_hosts(hostnames: &[String]) -> Result<()> {
    if hostnames.is_empty() {
        return Ok(());
    }
    let entries: Vec<String> = hostnames.iter()
        .map(|h| format!("127.0.0.1 {h}"))
        .collect();
    let cmd = format!("echo '{}' >> /etc/hosts", entries.join("\n"));
    let status = Command::new("sudo")
        .args(["sh", "-c", &cmd])
        .status()?;
    if !status.success() {
        bail!("failed to update /etc/hosts (sudo required)");
    }
    Ok(())
}

File: internal/services/ip.go
package services

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

const (
	networkStateFile = ".tncli/network.json"
	currentVersion   = 2
)

type NetworkState struct {
	Version     int               `json:"version"`
	Subnets     map[string]int    `json:"subnets"`
	Allocations map[string]string `json:"allocations"`
}

func homePath(rel string) string {
	home, _ := os.UserHomeDir()
	if home == "" {
		home = "/tmp"
	}
	return filepath.Join(home, rel)
}

func statePath() string { return homePath(networkStateFile) }

func LoadNetworkState() NetworkState {
	data, err := os.ReadFile(statePath())
	if err != nil {
		return NetworkState{Version: currentVersion, Subnets: make(map[string]int), Allocations: make(map[string]string)}
	}
	var state NetworkState
	if err := json.Unmarshal(data, &state); err != nil {
		return NetworkState{Version: currentVersion, Subnets: make(map[string]int), Allocations: make(map[string]string)}
	}
	if state.Subnets == nil {
		state.Subnets = make(map[string]int)
	}
	if state.Allocations == nil {
		state.Allocations = make(map[string]string)
	}
	return state
}

func saveNetworkState(state *NetworkState) {
	path := statePath()
	_ = os.MkdirAll(filepath.Dir(path), 0o755)
	data, err := json.MarshalIndent(state, "", "  ")
	if err == nil {
		_ = os.WriteFile(path, data, 0o644)
	}
}

func LoadIPAllocations() map[string]string {
	return LoadNetworkState().Allocations
}

// WithIPLock provides file-lock protected operation.
func WithIPLock(fn func()) {
	lockPath := homePath(".tncli/network.lock")
	_ = os.MkdirAll(filepath.Dir(lockPath), 0o755)

	for {
		f, err := os.OpenFile(lockPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o644)
		if err == nil {
			fmt.Fprintf(f, "%d", os.Getpid())
			f.Close()
			break
		}
		// Check stale lock
		if info, err := os.Stat(lockPath); err == nil {
			if time.Since(info.ModTime()) > 10*time.Second {
				_ = os.Remove(lockPath)
				continue
			}
		}
		time.Sleep(50 * time.Millisecond)
	}

	fn()
	_ = os.Remove(lockPath)
}

// MigrateLegacyIPs migrates from v1 to v2 format.
func MigrateLegacyIPs() {
	state := LoadNetworkState()
	if state.Version >= currentVersion {
		return
	}

	WithIPLock(func() {
		state := LoadNetworkState()
		if state.Version >= currentVersion {
			return
		}

		// Import from old files
		oldLoopback := homePath(".tncli/loopback.json")
		if data, err := os.ReadFile(oldLoopback); err == nil {
			var allocs map[string]string
			if json.Unmarshal(data, &allocs) == nil {
				for k, v := range allocs {
					if !strings.HasPrefix(v, "127.0.0.") {
						state.Allocations[k] = v
					}
				}
			}
			_ = os.Remove(oldLoopback)
		}

		oldSubnets := homePath(".tncli/subnets.json")
		if data, err := os.ReadFile(oldSubnets); err == nil {
			var subs map[string]int
			if json.Unmarshal(data, &subs) == nil {
				state.Subnets = subs
			}
			_ = os.Remove(oldSubnets)
		}

		// Clear legacy 127.0.0.x
		for k, ip := range state.Allocations {
			if strings.HasPrefix(ip, "127.0.0.") {
				delete(state.Allocations, k)
			}
		}

		// Clear old proxy routes
		routes := LoadRoutes()
		changed := false
		for k, target := range routes.Routes {
			if strings.HasPrefix(target, "127.0.0.") {
				delete(routes.Routes, k)
				changed = true
			}
		}
		if changed {
			recalcListenPorts(&routes)
			SaveRoutes(&routes)
		}

		state.Version = currentVersion
		saveNetworkState(&state)
		_ = os.Remove(homePath(".tncli/.migrated-subnet"))
	})
}

// MainIP gets IP for main workspace.
func MainIP(session, defaultBranch string) string {
	key := "ws-" + strings.ReplaceAll(defaultBranch, "/", "-")
	return AllocateIP(session, key)
}

// AllocateIP allocates next available loopback IP within the session's subnet.
// Format: 127.0.{subnet}.{2..254} — one subnet per session, one IP per workspace.
// Thread-safe via file lock. Ensures loopback alias exists after allocation.
func AllocateIP(session, worktreeKey string) string {
	var result string
	WithIPLock(func() {
		state := LoadNetworkState()

		// Each session gets its own /24 subnet (127.0.{N}.0/24)
		subnet, ok := state.Subnets[session]
		if !ok {
			used := make(map[int]bool)
			for _, s := range state.Subnets {
				used[s] = true
			}
			for n := 1; n <= 254; n++ {
				if !used[n] {
					subnet = n
					break
				}
			}
			if subnet == 0 {
				subnet = 254
			}
			state.Subnets[session] = subnet
		}

		// Return existing allocation
		if ip, ok := state.Allocations[worktreeKey]; ok {
			result = ip
			return
		}

		// Pick next available IP in subnet
		prefix := fmt.Sprintf("127.0.%d.", subnet)
		used := make(map[string]bool)
		for _, ip := range state.Allocations {
			used[ip] = true
		}
		for n := 2; n < 255; n++ {
			ip := fmt.Sprintf("%s%d", prefix, n)
			if !used[ip] {
				state.Allocations[worktreeKey] = ip
				saveNetworkState(&state)
				result = ip
				return
			}
		}

		// Subnet full — use last IP as fallback
		fallback := fmt.Sprintf("%s254", prefix)
		state.Allocations[worktreeKey] = fallback
		saveNetworkState(&state)
		result = fallback
	})

	// Ensure loopback alias exists (outside lock — calls daemon via socket)
	if result != "" {
		ensureLoopbackAlias(result)
	}
	return result
}

// ensureLoopbackAlias creates a loopback alias via the loopback daemon (no sudo needed).
// Falls back to sudo -n if daemon is not running.
func ensureLoopbackAlias(ip string) {
	if exec.Command("ping", "-c", "1", "-W", "1", ip).Run() == nil {
		return
	}
	// Ask daemon (runs as root via LaunchDaemon — no sudo prompt)
	if RequestLoopbackAlias(ip) {
		return
	}
	// Fallback: try non-interactive sudo
	_ = exec.Command("sudo", "-n", "ifconfig", "lo0", "alias", ip).Run()
}

// ReleaseIP releases an allocated IP.
func ReleaseIP(worktreeKey string) {
	WithIPLock(func() {
		state := LoadNetworkState()
		delete(state.Allocations, worktreeKey)
		saveNetworkState(&state)
	})
}

// CheckEtcHosts returns hostnames missing from /etc/hosts.
func CheckEtcHosts(hostnames []string) []string {
	content, _ := os.ReadFile("/etc/hosts")
	contentStr := string(content)
	var missing []string
	for _, h := range hostnames {
		if !strings.Contains(contentStr, h) {
			missing = append(missing, h)
		}
	}
	return missing
}
